Adding @font-feature-values

KenjiBaheux opened the pull request:
Trying to add @font-feature-values to the spec based on insights gained from experimenting with the @font-face/font-display descriptor and third party served web fonts.

Tab, does this look reasonable?
Should I expand with examples and details about what this would mean for the user agent?
Examples:

```
<link rel="https://fonts.googleapis.com/css?family=Roboto" rel="stylesheet">
<style>
@font-feature-values Roboto {
 font-display: optional;
}
</style>
```

Optional is set on a font-family with no refinement in individual font-face.
- The UA only use the font family if _all_ the required font-face are readily available
- otherwise the UA would just use the fallback font.
# 

```
<style>
@font-face {
  font-family: 'Roboto';
  font-style: normal;
  font-weight: 400;
  src: local('Roboto'), local('Roboto-Regular'), url(/roboto/regular.woff2) format('woff2');
}

@font-face {
  font-family: 'Roboto';
  font-style: normal;
  font-weight: 700;
  src: local('Roboto Bold'), local('Roboto-Bold'), url(/roboto/bold.woff2) format('woff2');
}

@font-face {
  font-family: 'Roboto';
  font-display: block;
  font-style: italic;
  font-weight: 400;
  src: local('Roboto Italic'), local('Roboto-Italic'), url(/roboto/italic.woff2) format('woff2');
}

@font-feature-values Roboto {
 font-display: fallback;
}
</style>
```

font-display: fallback is set on a font-family with refinement in some individual font-face rules:
-  The UA only use the font family if the regular and bold font-face are quickly available, otherwise the UA would just start with the fallback font and stick with it if too much time passes.
- On the other hand, the UA will block on the italic font face because its font-display descriptor trumps the default set by @font-feature-values.

KenjiBaheux committed:
Adding @font-feature-values

Trying to add @font-feature-values to the spec based on insights gained from experimenting with the @font-face/font-display descriptor and third party served web fonts.

Tab, does this look reasonable?
Should I expand with examples and details about what this would mean for the user agent?
Examples:
''
<link rel="https://fonts.googleapis.com/css?family=Roboto" rel="stylesheet">
<style>
@font-feature-values Roboto {
 font-display: optional;
}
</style>
''
Optional is set on a font-family with no refinement in individual font-face.
 - The UA only use the font family if *all* the required font-face are readily available
 - otherwise the UA would just use the fallback font.

''
<style>
@font-face {
  font-family: 'Roboto';
  font-style: normal;
  font-weight: 400;
  src: local('Roboto'), local('Roboto-Regular'), url(/roboto/regular.woff2) format('woff2');
}

@font-face {
  font-family: 'Roboto';
  font-style: normal;
  font-weight: 700;
  src: local('Roboto Bold'), local('Roboto-Bold'), url(/roboto/bold.woff2) format('woff2');
}

@font-face {
  font-family: 'Roboto';
  font-display: block;
  font-style: italic;
  font-weight: 400;
  src: local('Roboto Italic'), local('Roboto-Italic'), url(/roboto/italic.woff2) format('woff2');
}

@font-feature-values Roboto {
 font-display: fallback;
}
</style>
''

2. fallback set on a font-family with refinement in some individual font-face rules:
  -  The UA only use the font family if the regular and bold font-face are quickly available, otherwise the UA would just start with the fallback font and stick with it if too much time passes.
  - On the other hand, the UA will block on the italic font face because its font-display descriptor trumps the default set by @font-feature-values.

igrigorik commented on css-font-display/index.bs:
```diff
@@ -276,6 +277,21 @@ for example, forcing all fonts to have a ''0s'' <a>block period</a>.
+===========================
+The '@font-feature-values/font-display' descriptor for ''@font-feature-values''
+determines how a font family is displayed, by setting the "default" font-display value for @font-face rules targeting the same font family.
+In other words, @font-feature-values/font-display will set the value of "font-display:auto" (and its omission) for the @font-face rules of a given font-family.
```
Perhaps we can collapse these two? 

"targeting the same font family. If omitted, the default value of "font-display: auto" will be used by the user agent".

KenjiBaheux replied:
How about:

The '@font-feature-values/font-display' descriptor for ''@font-feature-values'' determines how a font family is displayed, by setting the "default" font-display value for @font-face rules targeting the same font family. When font-display is omitted in an @font-face rule, the user agent uses "font-display: auto" and therefore the "default" font-display value set via the @font-feature-values/font-display for the relevant font-family, if any.

?

igrigorik replied:
The last bit sounds a bit confusing to me. 

"When font-display is omitted in an @font-face rule, the user agent uses the font-display value set via the @font-feature-values/font-display for the relevant font-family if one is set, and otherwise defaults to "font-display: auto"."

igrigorik commented on css-font-display/index.bs:
```diff
@@ -276,6 +277,21 @@ for example, forcing all fonts to have a ''0s'' <a>block period</a>.
+determines how a font family is displayed, by setting the "default" font-display value for @font-face rules targeting the same font family.
+In other words, @font-feature-values/font-display will set the value of "font-display:auto" (and its omission) for the @font-face rules of a given font-family.
+
+This can be used to avoid the ransom note effect (i.e. "different loading behavior on different parts of the page") and to customize third party served fonts for which the modification of @font-face rules is not always possible.
```
I think its worth being a little more explicit here: (1) explain what problem this solves, (2) explain why it's applied to entire font-family.

This mechanism can be used to set a default rendering policy for the entire font-family, and enables developers to set a rendering policy for @font-face rules that are not directly under their control. For example, when a font is served by a third-party font foundry, the developer does not control the @font-face rules but is still able to set a default font-display policy for the  provided font-family. The policy is applied to the entire font family to avoid the ransom note effect (i.e. mismatched font faces). 

</bikeshed> feel free to ignore :)

KenjiBaheux replied:
How about:

This mechanism can be used to set a default display policy[1] for an[2] entire font-family, and enables developers to set a display policy for @font-face rules that are not directly under their control[3]. For example, when a font is served by a third-party font foundry, the developer does not control the @font-face rules but is still able to set a default font-display policy for the provided font-family. The ability to set a default policy for an entire font-family is also useful to avoid the ransom note effect (i.e. mismatched font faces) because the display policy is then applied to the entire font family.[4]

Notes:
- [1] The spec uses "font display strategy" and "font rendering strategy", we should probably converge on one. I like your use of "policy". Given that this used to be called font-rendering but then renamed to font-display, I'm thinking that we should use the term "font display policy" / "display policy".
- [2] "the => an" nit, not a native speaker but this feels better to me. Maybe I'm wrong :)
- [3] wondering about the following, suppose that you are using a third party who did include font-display values in his font-face rules. Can you redefine those rules? I'm assuming no because the font-face rules are more specific and therefore trump the @font-feature-values rule. Not sure if this is an important use case.
- [4] reinforcing the idea that there is a second use case

Thoughts?

igrigorik replied:
(1) +1 on converging on policy. 
(3) yes, that's my understanding too. I think this is fine and I assume this is a consistent behavior with other properties set by font-feature-values?
(4) Maybe we should move this into a non-normative note? Either that, or skip it entirely?

👍

igrigorik commented:
👍 👍 👍 

A few "bikeshed" nits above. 

I think font-family default is OK. It may be worth noting that the developer _could_ define different font-families that are backed by variants of the same font and apply different font-display policies to each.

KenjiBaheux committed:
Addressing Ilya's feedback

1. rephrasing the first paragraph of @font-feature-values/font-display
2. reworked second paragraph for @font-feature-values/font-display to: (1) explain what problem this solves, (2) explain why it's applied to entire font-family.
3. converging on "font display policy" across the spec (before: a mix of font rendering strategy and font display strategy)

KenjiBaheux committed:
Merge pull request #2 from KenjiBaheux/KenjiBaheux-fontfeaturevalues-patch-1

Addressing Ilya's feedback

KenjiBaheux commented:
Updated to address Ilya's feedback
